### DISCUSSION TITLE
Dynamically set schema

### DISCUSSION THREAD

**Discussion by wdfinch** [react-hook-form/react-hook-form#3972] [Ideas]
I had an idea for a feature based on one of my use cases. What do you all think of dynamically being able to change the scheme (yup in my case) of a `useForm`? So for example a method call `setScheme` that could change out a schema at any point. 

Use case:
I have a parent component that can be passed a child component. The child component has a form in it and the parent doesn't know what child it will receive. The parent needs to know if the child form is valid and manipulate it. Rather than use custom context and repeat logic to sync in all the child forms, it's easier to have the `useForm` hook in the parent and use the useFromContext in the children to communicate. This does mean that the child would choose the schema rather than the parent for validation and it could change as new children are mounted. Would be cool if I could change out the schema easily when a new child form is mounted.

**Comment by bluebill1049** [comment]
Thanks @wdfinch.I think you can pretty sure do that at the resolver now and pass down the context to switch the schema within. is this what you refer to?

**Reply by wdfinch** [reply]
I think so. I'll try that out. Thanks for the tip.

**Reply by charleshimmer** [reply]
@wdfinch what you ended up doing? We are migrating from Formik to react hook form, and with formik had a similar setup where children components (we called them SubForms) could register their yup schema with the parent form's schema. 

@bluebill1049 I'm using the yup resolver from here: https://github.com/react-hook-form/resolvers. How would we pass the context down?

**Reply by RasenGUY** [reply]
is this already resolved ? im trying to do something similar to wdfinch, im also wondering how to pass the context down

**Reply by yudielcurbelo** [reply]
A `setScheme` method will be really helpful to be able to change the schema from  `useFormContext`

**Reply by bombillazo** [reply]
Im able to do this via a setState value in the component I use `useForm` from, setting the validationSchema value here and passing the setter to my Form components where I need to update the validationSchema from.

I was wondering if there was a way to set the schema _after_ the useForm hook?

All this would be much more streamlined if RHF provided a `setValidationSchema` function one could call from the useFormContext/useForm hooks.

**Reply by ashrafnazar** [reply]
Has anyone come up with a solution for this? If so, a working example would be reaaaaally helpful 🙏

**Comment by chamatt** [comment]
+1 on this. Same use case, we have subforms (sections) inside the main form with conditionally rendered sections. I'd like to place the validation for a section inside the section itself, without piling up everything inside the parent component which makes everything too cumbersome to maintain.

A setValidationSchema/updateValidationSchema would be an amazing edition. I'm working on building a custom implementation for my company, if successful I'll share the results here.

**Reply by bluebill1049** [reply]
please do share or even publish a package if it's reusaeable.

**Reply by chamatt** [reply]
Here's the initial working result with a decent amount of type-safety (only works if using yup-resolver, but anyone can extend this if they want).

It adds a new `updateSchema` method to useForm that takes in a new Yup schema and merges it with the one from the initial hook call. In my use-case, for example, I'd call this in a `useEffect` when the section initially renders.
One small but important change is that you need to pass in a `schema` instead of a `resolver` when using the hook. There's an example usage after the code.

```typescript
/* eslint-disable @typescript-eslint/no-explicit-any */
import { useCallback, useState } from "react"
import { FieldValues, useForm, UseFormProps, UseFormReturn } from "react-hook-form"
import * as Yup from "yup"
import { yupResolver } from "@hookform/resolvers/yup"

type WidenSchema<T> = T extends (...args: any[]) => any
  ? T
  : { [K in keyof T]: T[K] extends Record<string, unknown> ? WidenSchema<T[K]> : any }

export type useCustomFormSchema<TFieldValues> = Yup.ObjectSchema<
  Yup.Shape<object | undefined, Partial<WidenSchema<TFieldValues>>>,
  object
>

export type UseCustomFormReturn<
  TFieldValues extends FieldValues = FieldValues,
  TContext = any,
> = UseFormReturn<TFieldValues, TContext> & {
  updateSchema: (newSchema: useCustomFormSchema<TFieldValues>) => void
}

/**
 * A custom useForm hook that allows dynamic validation schemas
 * @example
 * // Let's say we have a form with fields described by the following interface
 * interface FormValues {
 *  name: string
 *  age: number
 * }
 *
 * // And we setup the form with the following schema (only validating the name)
 * const { updateSchema, ...otherMethods } = useCustomForm<FormValues>({
 *  schema: Yup.object().shape({})
 * })
 *
 * // We can then update the schema to additionally validate the age field by calling
 * // the updateSchema method somewhere in the application
 * updateSchema(Yup.object().shape({ age: Yup.number().required() }))
 *
 * // The final validation schema will be
 * Yup.object().shape({
 *   name: Yup.string().required(),
 *   age: Yup.number().required()
 * })
 */
export const useCustomForm = <TFieldValues extends FieldValues = FieldValues, TContext = any>(
  props?: Omit<UseFormProps<TFieldValues, TContext>, "resolver"> & {
    schema?: useCustomFormSchema<TFieldValues>
  },
): UseCustomFormReturn<TFieldValues> => {
  const [schema, setSchema] = useState<useCustomFormSchema<TFieldValues> | null>(
    props?.schema ?? null,
  )

  const updateSchema = useCallback((newSchema: useCustomFormSchema<TFieldValues>) => {
    setSchema(oldSchema => {
      if (oldSchema) {
        return Yup.object().shape({
          ...oldSchema.fields,
          ...newSchema.fields,
        }) as useCustomFormSchema<TFieldValues>
      }
      return newSchema
    })
  }, [])

  const form = useForm({
    ...props,
    resolver: schema ? yupResolver<useCustomFormSchema<TFieldValues>>(schema) : undefined,
  })

  return {
    ...form,
    updateSchema,
  }
}

```
Example usage:
```typescript
// Let's say we have a form with fields described by the following interface
interface FormValues {
 name: string
 age: number
}

// And we setup the form with the following schema (only validating the name)
const { updateSchema, ...otherMethods } = useCustomForm<FormValues>({
 schema: Yup.object().shape({ name: Yup.string().required() }),
})

// We can then update the schema to additionally validate the age field by calling
// the updateSchema method somewhere in the application
updateSchema(Yup.object().shape({ age: Yup.number().required() }))

// The final validation schema will be
Yup.object().shape({
  name: Yup.string().required(),
  age: Yup.number().required()
})
```

**Edit 1:** I changed the way the merge is done, instead of using concat on the schema, I'm spreading the fields of both the old and the new schema. This guarantees that the new schema always wins when there are fields with the same key on both. This allows you to make, for example, a *required* field *unrequired*, which wasn't possible with concat

**Reply by ehudkaldor** [reply]
hi all,
@bluebill1049 , any update on this making it into the code? this is exactly what i came here looking for - i am building a form from blocks, and i want each block to be able to add its own resolver (schema?) to the form, either with a default one or with one provided by the block's props. it is pretty cool i can use a Controller in the block thanks to useFormContext, which makes it more annoying that i cannot do the same for the resolver :)

**Reply by ggmartins091** [reply]
I'm trying to do something similar, I have a useFieldArray, and each item on the array can have different schemas, I use the formContext for this, but there is no way for me to add custom validation to theese items.

what chamatt proposed looks very nice, I'll probably have tweak something to be able to change the schema for a specific index of my fieldArray, but this is nice.

**Reply by chamatt** [reply]
After months of usage of the code above I'd recommend against using it, at least in the current form. It is a bit brittle and can easily lead to infinite-rerendering if used wrong. I also couldn't get it to work when running tests with jest for some reason. Another part that isn't great for me is that it hides the validation from the parent, it kind of just inserts itself up there, which may cause unexpected behavior if you don't know the children has a built-in validation (like when using someone else's code)

Now-days I'm using another strategy, which is to make the component define it's own partial validation schema and export it,  then I import it in the parent and assemble it into a complete schema.

If you establish some patterns it can get quite straightforward, and the fact that you visibly assemble it in the parent, makes it clear that a validation is present (updateSchema kinda hides it)

Example:
#### FormSectionOne.tsx
```tsx
type FormSectionOneFormValues = {
  formSectionOne?: {
    // ...
  }
}

FormSectionOne = (parentForm: UseFormReturn<T & FormSectionOneFormValues>) => {
  const form = narrowForm<FormSectionOneFormValues>(parentForm)
  
  return (...)
}

FormSectionOne.getValidationSchema = (t: TFunction) =>  Yup.object().shape<FormSectionOneFormValues>({ 
     formSectionOne: Yup.object().shape({ ... }) 
})
```
#### ParentComponent.tsx
```tsx
type ParentComponentFormValues = { example1: string } & FormSectionOneFormValues

const ParentComponent = () => {
  const validationSchema = useMemo(() => ParentComponent.getValidationSchema(t), [t])

  const form = useForm<ParentComponentFormValues>({
    resolver: yupResolver(validationSchema),
  })
  
  return (...)
}

ParentComponent.getValidationSchema = (
  t: TFunction,
): Yup.Schema<ParentComponentFormValues> => Yup.object()
    .shape({
        example1: Yup.string().required(),
        ...FormSectionOne.getValidationSchema(t).fields,  // <-- important part
    }).required()
```

**Reply by carolineartz** [reply]
@chamatt thx this is helpful for what I'm trying to do right now. What does `narrowForm` look like?

**Reply by chamatt** [reply]
Glad it was helpful to you! Here's more details: https://github.com/orgs/react-hook-form/discussions/9345#discussioncomment-4132235

**Comment by AhmedBaset** [comment]
[outdated see the update below]

~It appears that form options are mutable~

https://github.com/react-hook-form/react-hook-form/blob/61524aeb83d78e8ffc34221b1db638582d4fbff4/src/logic/createFormControl.ts#L1420-L1425

~Then we can do:~

```tsx
const { control } = useFormContext<T>();
// ...
if (someCondition) {
    // (optional) clearErrors() and revalidate after setting the new schema
	control._options = {
	  resolver: zodResolver(createZodSchema({ type, action })),
	}
}
```

## Update: 

A better way I ended up using, is the proposed by the maintainer 
> I think you can pretty sure do that at the resolver now and pass down the context to switch the schema within.

```tsx
const [context, setContext] = useState(...)
const form = useForm<Type>({
    context: context,
    resolver: (values, context: typeof context, options) => {
      // you can return the schema based on context.
      // for my use case, I can depend on formValues
      const isSomeCondition = values.someValue === someValue;
      const createResolver = zodResolver(
        getOrderSchema({
          type: isSomeCondition ? "some-value" : "some-value-else",
        })
      );
      return createResolver(values, context, options);
    },
    // ...
});
```

Then, when RHF needs to validate it will pass formValues, context, and some options

https://github.com/react-hook-form/react-hook-form/blob/61524aeb83d78e8ffc34221b1db638582d4fbff4/src/logic/createFormControl.ts#L405-L419

**Reply by homocodian** [reply]
what's the context here in Updated section?

**Reply by AhmedBaset** [reply]
Any value you pass to useForm

```js
useForm({
  context: { ... },
})
```

**Reply by ashrafnazar** [reply]
Is there a working example?

**Reply by Narixius** [reply]
@ashrafnazar This is how I came across this situation

```tsx

const baseSchema = z.object({
  ...
})
const specialSchema = z.object({
  ...
})

const MyComponent = () => {
  const schema = useRef(baseSchema);
  const form = useForm({
    context: {
      schema
    },
    resolver: (...opts) => {
      return zodResolver(opts[1].schema.current)(...opts).finally(() => {
        // reset to default schema
        schema.current = baseSchema;
      })
    }
  })
  const save = () => { ... }

  const handleSubmit = () => {
    form.handleSubmit(save)();
  }

  const handleSpecialSubmit = () => {
    // right before handling submit, update the validation schema
    schema.current = specialSchema;
    form.handleSubmit(save)();
  }

  return <div> ....</div>
}
```
So here I have 2 different submit handlers, `handleSubmit` works with `baseSchema` and `handleSpecialSubmit` works with `specialSchema`.

**Reply by luidmidev** [reply]
thank you

**Reply by ashrafnazar** [reply]
@AhmedBaset - What is `getOrderSchema` from the example you used? Is it native to your project?

**Reply by homocodian** [reply]
> @AhmedBaset - What is `getOrderSchema` from the example you used? Is it native to your project?

Just a simple function which returns a different order schema based on whether an order is already present (editing existing order as some fields are optional and read-only when when order is being edited) or a new order is being created.

**Reply by ashrafnazar** [reply]
Do we need to use state to update the context? Or can we just pass in the schema? I.e. if we know that schema A has property 'x' which doesn't exist in schema B, can we then pass the outcome? E.g.

```tsx
resolver: (values, context, options) => {
  const getSchema = 'x' in values ? schemaA : schemaB
  const createResolver = zodResolver(getSchema)

  return createResolver(values, context, options)
},
```

**Comment by jmayergit** [comment]
Based on @AhmedBaset 's answer I had success with the following

For context I'm using
- zodResolver
- validate mode submission
- revalidate mode submission

I determine which zod schema to use based on form data

```Typescript
  const methods = useForm({
    resolver: (data, context, options) => {
      if (data.something === "...") {
        return zodResolver(aSchema)(data, context, options);
      }

      if (data.somethingElse === "...") {
        return zodResolver(anotherSchema)(data, context, options);
      }

      return zodResolver(yetAnotherSchema)(data, context, options);
    },
  });
```

[optional] when typing data in the submit handler I pass a type of the merged zod schemas
[optional] with the above , I create default values that contain all properties of all schemas

This works really nicely

**Update:** I use a **permissive zod schema which includes all form fields** , this allows me to pass a **complete set of default values** without TS complaining , then as in the code above (aSchema, anotherSchema, etc.) I have **several schemas which extend the permissive schema** , inside these the actual validation occurs .

For example if you have two fields whose default values are an empty string the permissive schema may simply covers those with `z.string()` , then when you dynamically validate at your chose timing (submit, etc.) you choose a more specific schema that applies the validation you want , for the field(s) you want (e.g., `z.string()` -> `z.string().min(3)` , this can be literally anything) . 

It would be nice if you only receive the fields you care about but it's not too much trouble to manually pick those out .

**Reply by ashrafnazar** [reply]
@bluebill1049 - maybe this could make its way into the documentation? Will def improve DX
